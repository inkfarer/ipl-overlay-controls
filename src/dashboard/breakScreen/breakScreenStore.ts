import { ActiveBreakScene, MainFlavorText, NextRoundStartTime, RuntimeConfig } from 'schemas';
import { defineStore } from 'pinia';
import { BundleDeclaredConfig } from 'types/schemas/bundleDeclaredConfig';

const mainFlavorText = nodecg.Replicant<MainFlavorText>('mainFlavorText');
const nextRoundStartTime = nodecg.Replicant<NextRoundStartTime>('nextRoundStartTime');
const activeBreakScene = nodecg.Replicant<ActiveBreakScene>('activeBreakScene');
const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');
const bundleDeclaredConfig = nodecg.Replicant<BundleDeclaredConfig>('bundleDeclaredConfig', { persistent: false });

export const breakScreenReps = [
    mainFlavorText,
    nextRoundStartTime,
    activeBreakScene,
    runtimeConfig,
    bundleDeclaredConfig
];

export interface BreakScreenStore {
    mainFlavorText: MainFlavorText;
    nextRoundStartTime: NextRoundStartTime;
    activeBreakScene: ActiveBreakScene;
    runtimeConfig: RuntimeConfig;
    bundleDeclaredConfig: BundleDeclaredConfig;
}

export const useBreakScreenStore = defineStore('breakScreen', {
    state: () => ({
        mainFlavorText: null,
        nextRoundStartTime: { startTime: null, isVisible: null },
        activeBreakScene: null,
        runtimeConfig: null,
        bundleDeclaredConfig: null
    } as BreakScreenStore),
    actions: {
        setActiveBreakScene(newValue: ActiveBreakScene): void {
            activeBreakScene.value = newValue;
        },
        setMainFlavorText(newValue: string): void {
            mainFlavorText.value = newValue;
        },
        setNextRoundStartTimeVisible(newValue: boolean): void {
            nextRoundStartTime.value.isVisible = newValue;
        },
        setNextRoundStartTime(newValue: string): void {
            nextRoundStartTime.value.startTime = newValue;
        }
    }
});
