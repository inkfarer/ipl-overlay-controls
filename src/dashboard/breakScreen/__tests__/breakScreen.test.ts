import BreakScreen from '../breakScreen.vue';
import { config, mount } from '@vue/test-utils';
import { useBreakScreenStore } from '../breakScreenStore';
import { createTestingPinia, TestingPinia } from '@pinia/testing';
import { IplButton, IplInput } from '@iplsplatoon/vue-components';
import NextStageTimeInput from '../components/nextStageTimeInput.vue';

describe('BreakScreen', () => {
    let pinia: TestingPinia;

    config.global.stubs = {
        IplButton: true,
        IplInput: true,
        IplCheckbox: true,
        NextStageTimeInput: true,
        IplErrorDisplay: true,
        FontAwesomeIcon: true
    };

    beforeEach(() => {
        pinia = createTestingPinia();
        config.global.plugins = [pinia];

        const store = useBreakScreenStore();
        store.bundleDeclaredConfig = {};
        // @ts-ignore
        store.runtimeConfig = {
            activeGraphicsBundles: []
        };
    });

    it('matches snapshot without custom scenes', () => {
        const store = useBreakScreenStore();
        store.activeBreakScene = 'main';
        const wrapper = mount(BreakScreen);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches snapshot with custom scenes', () => {
        const store = useBreakScreenStore();
        store.activeBreakScene = 'casters';
        store.runtimeConfig.activeGraphicsBundles = ['sj-overlays', 'sj-overlays-addon'];
        store.bundleDeclaredConfig = {
            'sj-overlays': {
                scenes: [
                    { value: 'casters', names: { EN: 'Casters!' } }
                ]
            },
            'sj-overlays-addon': {
                scenes: [
                    { value: 'stats', names: { EN: 'Statistics' } },
                    { value: 'analysts', names: { EN: 'Analyst desk' } }
                ]
            },
            'disabled-bundle': {
                scenes: [
                    { value: 'cool-scene', names: { EN: 'This scene shouldn\'t be visible' } }
                ]
            }
        };

        const wrapper = mount(BreakScreen);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('disabled expected button if active scene is main', () => {
        const store = useBreakScreenStore();
        store.activeBreakScene = 'main';
        const wrapper = mount(BreakScreen);

        expect(wrapper.getComponent('[data-test="show-main-button"]').attributes().disabled).toEqual('true');
        expect(wrapper.getComponent('[data-test="show-teams-button"]').attributes().disabled).toEqual('false');
        expect(wrapper.getComponent('[data-test="show-stages-button"]').attributes().disabled).toEqual('false');
    });

    it('disabled expected button if active scene is teams', () => {
        const store = useBreakScreenStore();
        store.activeBreakScene = 'teams';
        const wrapper = mount(BreakScreen);

        expect(wrapper.getComponent('[data-test="show-main-button"]').attributes().disabled).toEqual('false');
        expect(wrapper.getComponent('[data-test="show-teams-button"]').attributes().disabled).toEqual('true');
        expect(wrapper.getComponent('[data-test="show-stages-button"]').attributes().disabled).toEqual('false');
    });

    it('disabled expected button if active scene is stages', () => {
        const store = useBreakScreenStore();
        store.activeBreakScene = 'stages';
        const wrapper = mount(BreakScreen);

        expect(wrapper.getComponent('[data-test="show-main-button"]').attributes().disabled).toEqual('false');
        expect(wrapper.getComponent('[data-test="show-teams-button"]').attributes().disabled).toEqual('false');
        expect(wrapper.getComponent('[data-test="show-stages-button"]').attributes().disabled).toEqual('true');
    });

    it('updates main flavor text on store change', async () => {
        const store = useBreakScreenStore();
        store.mainFlavorText = 'hello!!!';
        const wrapper = mount(BreakScreen);

        store.mainFlavorText = 'new text!!!';
        await wrapper.vm.$nextTick();

        expect(wrapper.getComponent('[name="break-main-flavor-text"]').attributes().modelvalue).toEqual('new text!!!');
    });

    it('does not update main flavor text on store change if input is focused', async () => {
        const store = useBreakScreenStore();
        store.mainFlavorText = 'hello!!!';
        const wrapper = mount(BreakScreen);
        const flavorTextInput = wrapper.getComponent<typeof IplInput>('[name="break-main-flavor-text"]');

        flavorTextInput.vm.$emit('focuschange', true);
        store.mainFlavorText = 'new text!!!';
        await wrapper.vm.$nextTick();

        expect(flavorTextInput.attributes().modelvalue).toEqual('hello!!!');
    });

    it('updates next round time on store change', async () => {
        const store = useBreakScreenStore();
        store.nextRoundStartTime = { startTime: '2020-05-12', isVisible: true };
        const wrapper = mount(BreakScreen);

        store.nextRoundStartTime.startTime = '2021-03-09';
        await wrapper.vm.$nextTick();

        expect(wrapper.getComponent('[data-test="next-stage-time-input"]').attributes().modelvalue).toEqual('2021-03-09');
    });

    it('does not update next round time on store change if input is focused', async () => {
        const store = useBreakScreenStore();
        store.nextRoundStartTime = { startTime: '2020-05-12', isVisible: true };
        const wrapper = mount(BreakScreen);
        const nextStageTimeInput = wrapper.getComponent<typeof NextStageTimeInput>('[data-test="next-stage-time-input"]');

        nextStageTimeInput.vm.$emit('focuschange', true);
        store.nextRoundStartTime.startTime = '2021-03-09';
        await wrapper.vm.$nextTick();

        expect(nextStageTimeInput.attributes().modelvalue).toEqual('2020-05-12');
    });

    it('updates data on main scene update button click', async () => {
        const store = useBreakScreenStore(pinia);
        store.setNextRoundStartTime = jest.fn();
        store.setMainFlavorText = jest.fn();
        store.mainFlavorText = 'flavor text??';
        store.nextRoundStartTime.startTime = 'start time!!';
        const wrapper = mount(BreakScreen);

        wrapper.getComponent<typeof IplInput>('[name="break-main-flavor-text"]').vm.$emit('update:modelValue', 'new text!!!');
        wrapper.getComponent<typeof NextStageTimeInput>('[data-test="next-stage-time-input"]').vm.$emit('update:modelValue', '2020-01-02');
        await wrapper.vm.$nextTick();

        wrapper.getComponent<typeof IplButton>('[data-test="update-main-scene-button"]').vm.$emit('click');

        expect(store.setMainFlavorText).toHaveBeenCalledWith('new text!!!');
        expect(store.setNextRoundStartTime).toHaveBeenCalledWith('2020-01-02');
    });

    it('reverts changes to main scene data on update button right click', async () => {
        const store = useBreakScreenStore();
        store.mainFlavorText = 'flavor text??';
        store.nextRoundStartTime.startTime = 'start time!!';
        const wrapper = mount(BreakScreen);

        wrapper.getComponent<typeof IplInput>('[name="break-main-flavor-text"]').vm.$emit('update:modelValue', 'new text!!!');
        wrapper.getComponent<typeof NextStageTimeInput>('[data-test="next-stage-time-input"]').vm.$emit('update:modelValue', '2020-01-02');
        wrapper.getComponent<typeof IplButton>('[data-test="update-main-scene-button"]').vm.$emit('rightClick', new Event(null));
        await wrapper.vm.$nextTick();

        expect(wrapper.getComponent('[data-test="next-stage-time-input"]').attributes().modelvalue).toEqual('start time!!');
        expect(wrapper.getComponent('[name="break-main-flavor-text"]').attributes().modelvalue).toEqual('flavor text??');
    });
});
