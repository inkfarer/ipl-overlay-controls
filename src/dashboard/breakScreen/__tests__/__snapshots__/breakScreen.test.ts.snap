// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreakScreen matches snapshot with custom scenes 1`] = `
"<ipl-error-display-stub class="m-b-8"></ipl-error-display-stub>
<div data-v-31c8f953="" class="ipl-expanding-space color-dark">
  <div data-v-31c8f953="" class="ipl-expanding-space__header">
    <div data-v-31c8f953="" class="ipl-expanding-space__header-background" tabindex="0"></div>
    <div data-v-31c8f953="" class="ipl-expanding-space__title">translation:sceneName.main</div>
    <font-awesome-icon-stub data-v-31c8f953="" icon="[object Object]" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false" class="icon"></font-awesome-icon-stub>
    <div data-v-31c8f953="" class="header-extra">
      <ipl-button-stub label="translation:showSceneButton" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="show-scene-button m-l-8" data-test="show-main-button"></ipl-button-stub>
    </div>
  </div>
  <div data-v-31c8f953="" class="content" style="display: none;">
    <ipl-input-stub name="break-main-flavor-text" label="translation:mainScene.flavorTextInput" type="text" centered="false" disabled="false" loading="false" theme="default"></ipl-input-stub>
    <div class="layout vertical center-horizontal">
      <div class="m-t-8">translation:nextStageTimer.sectionTitle</div>
      <next-stage-time-input-stub data-test="next-stage-time-input"></next-stage-time-input-stub>
      <ipl-checkbox-stub label="translation:nextStageTimer.showTimerCheckbox" small="false" disabled="false" class="m-t-8"></ipl-checkbox-stub>
      <ipl-button-stub label="common:button.update" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="m-t-8" data-test="update-main-scene-button" title="common:button.rightClickUndoMessage"></ipl-button-stub>
    </div>
  </div>
</div>
<div data-v-868771ad="" class="ipl-space color-dark layout horizontal center-vertical layout-break-scene m-t-8"><span>translation:sceneName.teams</span>
  <ipl-button-stub label="translation:showSceneButton" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="show-scene-button" data-test="show-teams-button"></ipl-button-stub>
</div>
<div data-v-868771ad="" class="ipl-space color-dark layout horizontal center-vertical layout-break-scene m-t-8"><span>translation:sceneName.stages</span>
  <ipl-button-stub label="translation:showSceneButton" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="show-scene-button" data-test="show-stages-button"></ipl-button-stub>
</div>
<div data-v-868771ad="" class="ipl-space color-dark layout horizontal center-vertical layout-break-scene m-t-8"><span>casters</span>
  <ipl-button-stub label="translation:showSceneButton" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="show-scene-button" data-test="show-scene-button__casters"></ipl-button-stub>
</div>
<div data-v-868771ad="" class="ipl-space color-dark layout horizontal center-vertical layout-break-scene m-t-8"><span>stats</span>
  <ipl-button-stub label="translation:showSceneButton" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="show-scene-button" data-test="show-scene-button__stats"></ipl-button-stub>
</div>
<div data-v-868771ad="" class="ipl-space color-dark layout horizontal center-vertical layout-break-scene m-t-8"><span>analysts</span>
  <ipl-button-stub label="translation:showSceneButton" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="show-scene-button" data-test="show-scene-button__analysts"></ipl-button-stub>
</div>
<a href="/dashboard/#workspace/ipl%20setup" target="_top"><button data-v-868771ad="" class="ipl-space color-dark extra-layout-button m-t-8">translation:sceneConfigLink</button></a>"
`;

exports[`BreakScreen matches snapshot without custom scenes 1`] = `
"<ipl-error-display-stub class="m-b-8"></ipl-error-display-stub>
<div data-v-31c8f953="" class="ipl-expanding-space color-dark">
  <div data-v-31c8f953="" class="ipl-expanding-space__header">
    <div data-v-31c8f953="" class="ipl-expanding-space__header-background" tabindex="0"></div>
    <div data-v-31c8f953="" class="ipl-expanding-space__title">translation:sceneName.main</div>
    <font-awesome-icon-stub data-v-31c8f953="" icon="[object Object]" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false" class="icon"></font-awesome-icon-stub>
    <div data-v-31c8f953="" class="header-extra">
      <ipl-button-stub label="translation:showSceneButton" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="show-scene-button m-l-8" data-test="show-main-button"></ipl-button-stub>
    </div>
  </div>
  <div data-v-31c8f953="" class="content" style="display: none;">
    <ipl-input-stub name="break-main-flavor-text" label="translation:mainScene.flavorTextInput" type="text" centered="false" disabled="false" loading="false" theme="default"></ipl-input-stub>
    <div class="layout vertical center-horizontal">
      <div class="m-t-8">translation:nextStageTimer.sectionTitle</div>
      <next-stage-time-input-stub data-test="next-stage-time-input"></next-stage-time-input-stub>
      <ipl-checkbox-stub label="translation:nextStageTimer.showTimerCheckbox" small="false" disabled="false" class="m-t-8"></ipl-checkbox-stub>
      <ipl-button-stub label="common:button.update" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="m-t-8" data-test="update-main-scene-button" title="common:button.rightClickUndoMessage"></ipl-button-stub>
    </div>
  </div>
</div>
<div data-v-868771ad="" class="ipl-space color-dark layout horizontal center-vertical layout-break-scene m-t-8"><span>translation:sceneName.teams</span>
  <ipl-button-stub label="translation:showSceneButton" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="show-scene-button" data-test="show-teams-button"></ipl-button-stub>
</div>
<div data-v-868771ad="" class="ipl-space color-dark layout horizontal center-vertical layout-break-scene m-t-8"><span>translation:sceneName.stages</span>
  <ipl-button-stub label="translation:showSceneButton" color="green" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" errormessage="Error!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" inline="false" class="show-scene-button" data-test="show-stages-button"></ipl-button-stub>
</div>
<!--v-if-->"
`;
